### PR TITLE
Rework of Webservice API

### LINKED ISSUE #22675: Request to get all product images
#### Is your feature request related to a problem?

Drugin synchronization of my system Products with Prestashop, Ineed to trigger request for each product to get product images. As a result, I am sending lots of requests like that:
https://presta:8888/api/images/products/1096
https://presta:8888/api/images/products/1097
... and so on
Thats very resource wasteful

#### Describe the solution you'd like

Simple request to get all of product images, forparsing later:
https://presta:8888/api/images/products
should return XML (or json) of all product images, where I can see which image is related to which product

#### Alternatives you've considered

Alternative is to get one product a time, which I said is very resource wasteful

### PR DESCRIPTION
**Pull request by KminekMatej**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | See below, its pretty huge
| Type?             | bug fix & improvement & new feature & refactor
| Category?         | WS
| BC breaks?        | yes
| Deprecations?     | no
| Fixed ticket?     | Fixes #22675
| How to test?      | Using new WS autotest
| Possible impacts? | Composing API has been completely refactored, modernized and enlighted. Every file underneath classes/webservice has been changed, some now added. There is just one actual impact, which has brought me to refactor API completely - getting product images. Please see diff in ([ https://www.diffchecker.com/fisRdJ1l ],) what has been actually changed. This should be the only BC break, contained in this PR, everything else should stay the same (as for now)

**Note**:
This PR replaces pull-requests #23100 and #27297

**Description**:

WebService API was having historical troubles. Its XML output has been using string composing of headers / footers (instead of using modern SimpleXML PHP interface), which could cause troubles (well actually didnt, at least in modern versions of Prestashop). Troubles started after JSON output has been added into Prestashop - you can't correctly compose JSON output using header and footer rendering. That led to some custom overriding functions, leading to a bit of chaos on API Core structure.

I tried to solve this design trouble by creating a class called ApiNode and refactored WebServiceOutputBuilder to firstly build structure of ApiNode tree. Each ApiNode is basically an object, composing of name, type, array of attributes and array of subnodes. This ApiNode is then processed by XML / JSON rendering classes.

Then, WebServiceOutputJSON and WebServiceOutputXML are using simple function renderNode(ApiNode), which iterates over ApiNode (and its children) and composes output string.

WebServiceOutputXML uses SuperXMLElement (which is actually SimpleXMLElement, just extended with addCData() function) to create XML hierarchy. Output is then properly XML formatted into string. That makes XML API output easily extendable in the future. XML output is no more pretty printed (using deprecated $depth property), to avoid unneccessary overhead (XML is usually pretty printed by modern browsers automatically anyway and there is no need for pretty printing in actual API calls)

WebServiceOutputJSON iterates over ApiNode and its children to create a simple array of objects and its subobjects, then transforms it into string, using json_encode common PHP function. Due to nature of JSON, we cannot put attributes in it, as in XML. Either we have to accept complete BC break and add attributes into JSON request as well, or avoid working with attributes and treat important attributes as children nodes, thus creating another BC break in XML output.
(Working with attributes will definitely be neccessary on images requests - but they havent been working anyway in JSON, so I believe there is no BC break). Non-BC possibility would thus be to add attributes as subnodes only on Images requests. But this solution needs further analyze.

Consider this PR as a draft. I am going to process further tests. I have already tested few requests (mainly on products), ensuring that XML and JSON outputs stays the same, but more tests are neccessary. I am not sure whether there are any API autotests, which would come very handy.

Also, code-review from original creators of Prestashop WebService API would be very handy. I think that during the code examination I discovered every potential use-case, but one can never be sure.

:notebook: **BC Breaks :**

1. Change property `WebserviceOutputBuilderCore::$wsUrl` to `public static`
2. Change input parameters on `WebserviceOutputBuilderCore::renderEntity`
3. Change input parameters on `WebserviceOutputBuilderCore::renderField`
4. Change input parameters on `WebserviceOutputBuilderCore::renderAssociations`
5. Removed 2nd param `$depth` on method `WebserviceOutputBuilderCore::renderEntityMinimum`
6. Removed method `WebserviceOutputBuilderCore::renderFlatAssociation`
7. Removed method `WebServiceOutputInterface::__construct`
8. Removed method `WebServiceOutputInterface::setWsUrl`
9. Removed method `WebServiceOutputInterface::getWsUrl`
10. Removed method `WebServiceOutputInterface::getContentType`
11. Removed method `WebServiceOutputInterface::setSchemaToDisplay`
12. Removed method `WebServiceOutputInterface::getSchemaToDisplay`
13. Removed method `WebServiceOutputInterface::renderField`
14. Removed method `WebServiceOutputInterface::renderNodeHeader`
15. Removed method `WebServiceOutputInterface::renderNodeFooter`
16. Removed method `WebServiceOutputInterface::renderAssociationHeader`
17. Removed method `WebServiceOutputInterface::renderAssociationFooter`
18. Removed method `WebServiceOutputInterface::overrideContent`
19. Removed method `WebServiceOutputInterface::renderErrorsHeader`
20. Removed method `WebServiceOutputInterface::renderErrorsFooter`
21. Removed method `WebServiceOutputInterface::renderErrors`
22. Removed method `WebServiceOutputInterface::setLanguages`
23. Removed method `WebServiceOutputInterface::renderAssociationWrapperHeader`
24. Removed method `WebServiceOutputInterface::renderAssociationWrapperFooter`
25. Add output type to method `WebServiceOutputInterface::getContentType`
26. Removed method `WebserviceOutputJSONCore::renderAssociationField`
27. Removed method `WebserviceOutputJSONCore::renderi18nField`
28. Removed property `WebserviceOutputJSONCore::$docUrl`
29. Removed property `WebserviceOutputJSONCore::$languages`
30. Removed property `WebserviceOutputJSONCore::$wsUrl`
31. Removed property `WebserviceOutputJSONCore::$schemaToDisplay`
32. Removed property `WebserviceOutputJSONCore::$currentEntity`
33. Removed property `WebserviceOutputJSONCore::$currentAssociatedEntity`
34. Removed property `WebserviceOutputJSONCore::$content`
35. Removed method `WebserviceOutputXMLCore::renderAssociationField`
36. Removed method `WebserviceOutputXMLCore::renderi18nField`
37. Removed property `WebserviceOutputXMLCore::$docUrl`
38. Removed property `WebserviceOutputXMLCore::$languages`
39. Removed property `WebserviceOutputXMLCore::$wsUrl`
40. Removed property `WebserviceOutputXMLCore::$schemaToDisplay`
41. Add input & output types to method `WebserviceSpecificManagementImages::manageDeclinatedImages`

 (BC breaks of interface WebServiceOutputInterface applies also to its implementors `WebserviceOutputJSONCore` and `WebserviceOutputXMLCore`)

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27297)

### CODE REVIEW

File: classes/webservice/ApiNode.php
Comment (sowbiba) [inline classes/webservice/ApiNode.php:342]: Can you describe in the phpDoc the keys that can exist in the $field array ?

File: classes/webservice/ApiNode.php
Comment (sowbiba) [inline classes/webservice/ApiNode.php:374]: In this block, you don't test if isset($field['value'])

**Review by sowbiba — COMMENTED** [review]
I just made the first pass of review
I like what you did. That's a great and huge work. Thanks for that

But 
- There is a lack of tests that will make the PR hard to accept
- Usually to replace a whole system by another, You can fully test the old one, replace by the new system without breaking the tests. OR you build the new system, make it functionnally tested by those who frequently use it, fix bugs and then make the switch

I'll let other reviewer appreciate the risk vs gain and tell how to confidently integrate your nice work

File: classes/webservice/ApiNode.php
Comment (KminekMatej) [reply classes/webservice/ApiNode.php:342]: Most of this logic is not changed and taken from previous api, which has not been documented at all :neutral_face: . I tried to describe the array inputs from what I found in code directly, please check 885dba5 whether its correct

File: classes/webservice/ApiNode.php
Comment (KminekMatej) [reply classes/webservice/ApiNode.php:374]: Added in accca16

**Comment by KminekMatej** [comment]
@sowbiba I am aware that this is a huge and very risky merge-request. It actually led @Progi1984 to create at least some autotests for a webservice part in last year. However, missing tests shouldnt be the reason to deny this upgrade, as I think its a great step forward to solve many issues that WS is currently having. 
Good way to go would be to extend code & functional coverage of tests to catch most of possible use-cases.

File: classes/webservice/ApiNode.php
Comment (Progi1984) [inline classes/webservice/ApiNode.php]:
```suggestion
 */

declare(strict_types=1);

class ApiNode
```

File: classes/webservice/SuperXMLElement.php
Comment (Progi1984) [inline classes/webservice/SuperXMLElement.php]:
```suggestion
 */

declare(strict_types=1);

class SuperXMLElement extends \SimpleXMLElement
```

File: classes/webservice/WebserviceOutputBuilder.php
Comment (Progi1984) [inline classes/webservice/WebserviceOutputBuilder.php]:
```suggestion
    public static $wsUrl;
    public static $languages;
```

Could you add PHPDoc for these attributes ?

File: classes/webservice/WebserviceOutputBuilder.php
Comment (Progi1984) [inline classes/webservice/WebserviceOutputBuilder.php:464]: Could you document the BC Break ?

File: classes/webservice/WebserviceOutputBuilder.php
Comment (Progi1984) [inline classes/webservice/WebserviceOutputBuilder.php:427]: Could you document the BC Break ?

File: classes/webservice/WebserviceOutputBuilder.php
Comment (Progi1984) [inline classes/webservice/WebserviceOutputBuilder.php:516]: Could you document the BC Break ?

File: classes/webservice/WebserviceOutputBuilder.php
Comment (Progi1984) [inline classes/webservice/WebserviceOutputBuilder.php:582]: Could you document the BC Break ?

File: classes/webservice/WebserviceOutputBuilder.php
Comment (Progi1984) [inline classes/webservice/WebserviceOutputBuilder.php:694]: Could you document the BC Break ? (Removal of this method)

File: classes/webservice/WebserviceOutputInterface.php
Comment (Progi1984) [inline classes/webservice/WebserviceOutputInterface.php:26]: Could you document all these BC Break ?

File: classes/webservice/WebserviceOutputInterface.php
Comment (Progi1984) [inline classes/webservice/WebserviceOutputInterface.php]:
```suggestion
```

File: classes/webservice/WebserviceOutputXML.php
Comment (Progi1984) [inline classes/webservice/WebserviceOutputXML.php]:
```suggestion
```

File: classes/webservice/WebserviceOutputXML.php
Comment (Progi1984) [inline classes/webservice/WebserviceOutputXML.php]:
```suggestion
    private function injectAttributes(SuperXMLElement &$xml, ApiNode $node): void
```

File: classes/webservice/WebserviceOutputXML.php
Comment (Progi1984) [inline classes/webservice/WebserviceOutputXML.php]:
```suggestion
    private function injectChildren(SuperXMLElement &$parentXml, ApiNode $apiNode): void
```

File: classes/webservice/WebserviceSpecificManagementSearch.php
Comment (Progi1984) [inline classes/webservice/WebserviceSpecificManagementSearch.php]:
```suggestion
```

**Review by Progi1984 — COMMENTED** [review]
@KminekMatej Good job. Thank you for your contribution. I have brought your some comments.

File: classes/webservice/ApiNode.php
Comment (KminekMatej) [reply classes/webservice/ApiNode.php]: Committed

File: classes/webservice/SuperXMLElement.php
Comment (KminekMatej) [reply classes/webservice/SuperXMLElement.php]: Committed

File: classes/webservice/WebserviceOutputBuilder.php
Comment (KminekMatej) [reply classes/webservice/WebserviceOutputBuilder.php]: Added in d3f57f3 - `$languages` has been removed since its unused here

File: classes/webservice/WebserviceOutputBuilder.php
Comment (KminekMatej) [reply classes/webservice/WebserviceOutputBuilder.php:464]: I dont think its a BC break, since I fixed all the internal calls of this function, no? Or why do you consider this as a BC break?

File: classes/webservice/WebserviceOutputBuilder.php
Comment (KminekMatej) [reply classes/webservice/WebserviceOutputBuilder.php:694]: If its not being called anywhere anymore, its not a BC break, no?

File: classes/webservice/WebserviceOutputXML.php
Comment (KminekMatej) [reply classes/webservice/WebserviceOutputXML.php]: Solved by 28826f0

File: classes/webservice/WebserviceOutputXML.php
Comment (KminekMatej) [reply classes/webservice/WebserviceOutputXML.php]: Solved by 28826f0

File: classes/webservice/WebserviceOutputBuilder.php
Comment (matthieu-rolland) [reply classes/webservice/WebserviceOutputBuilder.php:464]: Hello, changing the signature of a public method is considered a BC break, any third party element relying on this method would be broken

so even if its' unlikely, we document the BC break ;)

**Review by matthieu-rolland — COMMENTED** [review]
> Consider this PR as a draft. I am going to process further tests. I have already tested few requests (mainly on products), ensuring that XML and JSON outputs stays the same, but more tests are neccessary. I am not sure whether there are any API autotests, which would come very handy.

Maybe writing some kind of tests would help you see any misbehavior ? And it would also bring your PR closer to merge ;)

Anyway, thank you for the work, it's indeed a nice contribution 👍 

(seems like you have to rebase a fix a small conflict)

File: classes/webservice/ApiNode.php
Comment (Progi1984) [inline classes/webservice/ApiNode.php:38]:
```suggestion
    public static $languages = [];
```

Could you add the PHPDoc too ?

**Review by Progi1984 — COMMENTED** [review]
Good job.

Remains some feedbacks and document all BC Breaks, and it's good.

**Comment by KminekMatej** [comment]
@matthieu-rolland @Progi1984 Could you guys please direct me how to document the BC break? I cant find any relevant how-to in Dev docs

File: classes/webservice/ApiNode.php
Comment (KminekMatej) [reply classes/webservice/ApiNode.php:38]: #[c3e7481](https://github.com/PrestaShop/PrestaShop/pull/27366/commits/c3e74818147913101aaed021bce67c8168477c52)

**Comment by kpodemski** [comment]
@KminekMatej 

Here you have a PR that has information about BC Breaks in the description:
https://github.com/PrestaShop/PrestaShop/pull/28102

Later on, we'll add a proper label to make sure to copy this information to the dev docs.

File: classes/webservice/WebserviceOutputBuilder.php
Comment (KminekMatej) [reply classes/webservice/WebserviceOutputBuilder.php:464]: All BC breaks documented in description as advised. The list is pretty huge :)

**Comment by KminekMatej** [comment]
All BC breaks documented in description as advised. The list is pretty huge :)

**Comment by Progi1984** [comment]
@KminekMatej You can check how I define BC Breaks here : #27590.

**Review by Progi1984 — COMMENTED** [review]
@KminekMatej Could you rebase your PR, please ? You have some conflicting files.

**Comment by KminekMatej** [comment]
Rebased. Hope everything will be fine.
@Progi1984 My BC breaks docs looks alike yours, is that ok now?

**Review by justeen35 — DISMISSED** [review]

**Review by Progi1984 — APPROVED** [review]

**Review by Progi1984 — CHANGES_REQUESTED** [review]
@KminekMatej Could you fix behat tests & PHPCSFixer please ? And you will have my approval.

**Comment by KminekMatej** [comment]
@Progi1984 I found 2 errors and fixed them. Now all the tests are passing

**Review by Progi1984 — DISMISSED** [review]

**Comment by Xzeno15150** [comment]
Hello,

I tried to use what you did in a project, and it's a great job. 
I aimed to find url and the id of a specified product image. With the XML output it works fine, but with the JSON we don't get any of these.

With the XML format (http://_shop_/api/images/products/9&output_format=XML) :
image

With the JSON Format (http://_shop_/api/images/products/9&output_format=JSON) :
image

**Comment by KminekMatej** [comment]
> Hello,
> 
> I tried to use what you did in a project, and it's a great job. I aimed to find url and the id of a specified product image. With the XML output it works fine, but with the JSON we don't get any of these.
> 
> With the XML format (http://_shop_/api/images/products/9&output_format=XML) : image
> 
> With the JSON Format (http://_shop_/api/images/products/9&output_format=JSON) : image

Hello @Xzeno15150 ,
are you sure you have the latest pull? That certainly looks like the issue Ive been fixing 2 days ago .. But let me test that on my instance tonight and get back to you

**Comment by Xzeno15150** [comment]
> > Hello,
> > I tried to use what you did in a project, and it's a great job. I aimed to find url and the id of a specified product image. With the XML output it works fine, but with the JSON we don't get any of these.
> > With the XML format (http://_shop_/api/images/products/9&output_format=XML) : image
> > With the JSON Format (http://_shop_/api/images/products/9&output_format=JSON) : image
> 
> Hello @Xzeno15150 , are you sure you have the latest pull? That certainly looks like the issue Ive been fixing 2 days ago .. But let me test that on my instance tonight and get back to you

I just cloned and checkout to this PR to test if the problem wasn't me

**Comment by KminekMatej** [comment]
I found the problem and fixed it. However It seems to me that original behaviour was erroneous as well. Behold the original JSON response:
```
{
    "": [
        {
            "id": "9"
        },
        {
            "id": "10"
        }
    ]
}
```

After my fix, the response now looks like this:
```
{
    "image": [
        {
            "id": "9"
        },
        {
            "id": "10"
        }
    ]
}
```
Only difference is that I am putting name of parent node as an object. It almost looks like that the original renderer tried to do that as well, but somehow failed and put empty string instead.

Thats a breaking change, yet - however - I think its also a bug fix.
Please give me your opinion @matthieu-rolland @Progi1984 - should I keep this behaviour (with the `image` key) or should I rewrite that to keep the empty string as key?

**Comment by KminekMatej** [comment]
I rebased this to develop branch, due to merge conflicts

**Review by Progi1984 — APPROVED** [review]

File: classes/webservice/ApiNode.php
Comment (MeKeyCool) [inline classes/webservice/ApiNode.php:335]: Thanks a lot for this detailed documentation ^^

Maybe some indentation would make it easier to read ? :thinking: 
```php
    /*
    * ...
    *     <pre>
    *     $field = [
    *         'sqlId' => (string) DB field name. Required.
    *         'encode'    => (string) Field encoding. May contain 'base64' for displaying image content. Optional.
    *         'synopsis_details'  => [
    *         'required'  => (string) 'true' / 'false'. Added as an attribute.
    *         'maxSize'   => (int) Maximum field size. Length for strings, digits for integers. Added as an attribute.
    *         'format'    => (string) Field specific format. Added as an attribute.
    *         'readOnly'  => (string) 'true' / 'false'. Added as an attribute.
    *       ] (Optional)
    *       'i18n'  => (bool) If this is international field. If set, this field is rendered with <strong>language</strong> sub-nodes. (Optional)
    *       'value' => (mixed) Field value. Required.
    *       'xlink_resource'    => (string) Url resource. If set to string, this field is rendered with <strong>xlink:href</strong> attribute, containing this url, prepended with language base url.
    *       'xlink_resource'    => [
    *         'resourceName'  => (string) Resource name, to url scheme: <i>'wsUrl/<strong>resourceName</strong>/<strong>subResourceName</strong>/<strong>object_id</strong>/<strong>value</strong>'</i>
    *         'subResourceName'  => (string) Sub-Resource name, used to compose output url.
    *       ]
    *       'object_id' => (string|int) Sub resource id. Used to compose output url and only if xlink_resource contains subResourceName.
    *       'getter'    => (string) Name of getter. If set, adds attribute <strong>notFilterable = true</strong>.
    *       'setter'    => (bool) Name of setter. If set to <i>false</i> and <strong>$schemaToDisplay</strong> is set to synopsis, adds attribute <strong>read_only = true</strong>.
    *     ]
    *     </pre>
    * ...
    */
```

File: classes/webservice/ApiNode.php
Comment (MeKeyCool) [inline classes/webservice/ApiNode.php:376]: As `WebserviceOutputBuilderCore::$wsUrl` is a public static attribute in a normal class, how do you ensure that it is set at this moment ?

The problem i see is that this `public function addField(array $field, ?string $schemaToDisplay = null): self` is almost a factory but the context it uses to build new objects is global.

I suspect we encounter some troubles if we try to test this code.

File: classes/webservice/WebserviceOutputBuilder.php
Comment (MeKeyCool) [inline classes/webservice/WebserviceOutputBuilder.php:46]: As you refactored almost everything in this class, I am quite suspicious with this attribute usage.

As the state of this class is implied in a string construction, what would happen If we use the same builder for several strings ? :thinking: 

The usage workflow of this class is quite difficult to understand reading code from nowhere. Some basic test may help a lot to understand basic usage ^^

File: classes/webservice/ApiNode.php
Comment (KminekMatej) [reply classes/webservice/ApiNode.php:376]: Hello,
static property `$wsUrl` is actually replacement of class properties `WebserviceOutputBuilder::$wsUrl`, `WebserviceOutputJSON::$wsUrl` and `WebserviceOutputXML::$wsUrl`. Its being initialized frorm `WebserviceRequest->fetch(..)` method, which I believe is a starting point of each webservice request and then, during app flow, WebServiceOutputXX classes were being constructed using this `$wsUrl` string. Since `WebserviceOutputBuilder::$wsUrl` had been inited directly in constructor from `WebserviceRequest`, without any object-related logic, I made this variable static and declared on one place (`WebserviceOutputBuilder`) only, which made the code whole lot more readable (not mentioning slight performance gain).

It may seem like, but I really didnt rewrite the whole logic of PS WebService and actually didnt ask myself the question - how is ensured that this variable is always filled? I simply count that the one who wrote the `Webservice` was pretty sure :). Anyway, this variable is inited in `WebserviceOutputBuilder` constructor and looking at the constructing logic, guess I can be very sure that this variable is always inited.

File: classes/webservice/WebserviceOutputBuilder.php
Comment (KminekMatej) [reply classes/webservice/WebserviceOutputBuilder.php:46]: As far as I can see, the $wsUrl is constructed like this:
`Tools::getHttpHost(true) . __PS_BASE_URI__ . 'api/'`
That means that for one PS instance is static and unchangeable

Comment (KminekMatej) [reply classes/webservice/WebserviceOutputBuilder.php:46]: I totally agree that the usage of this class is quite difficult to understand. I spent several hours trying to decipher the code even before this change (which I still believe made the code lot easier to understand).

What would really be helpful to write a lot more tests. Not for this current PR, but for the current Webservice implementation. Yes, those tests may help a lot to understand basic usage. But since I am refactoring functionality, currently running on milions of places, tests for current webservice behaviour would be very handy, as they can serve as a borders to any webservice upgrades, including this one. Im almost sure those tests could reveal a lot of errors even in current WS

File: classes/webservice/WebserviceSpecificManagementImages.php
Comment (metacreo) [inline classes/webservice/WebserviceSpecificManagementImages.php:1]: `_PS_PRODUCT_IMG_DIR_ -> _PS_PROD_IMG_DIR_`

Comment (metacreo) [reply classes/webservice/WebserviceSpecificManagementImages.php:1]: `manageDeclinatedImages(string $directory) : bool`

Fatal.

**Comment by KminekMatej** [comment]
Currently Im not working on Prestashop API refactoring anymore as I comply with the statement that before API gets merged, we need to write whole bunch of API autotests.
However due to my different scope at work, I dont have the capacity nor knowledge needed to write such tests.

**Comment by vitalijalbu** [comment]
Would be better leave only JSON, nowdays no one is using XML anymore... or at least make JSON as default

**Comment by metacreo** [comment]
@vitalijalbu XML is used in many modules. For example FedEx carrier, DHL carrier. 
It is impossible to completely and immediately drop XML support.

**Comment by sneko** [comment]
Any plan from the core team to fix issues from the JSON API? I'm coming from the issue https://github.com/PrestaShop/PrestaShop/issues/15499 dated from 2019 and the different PR attempts from @KminekMatej were meaningful.

In the meantime I had a lot of hopes with Prestashop v9 with the new OpenAPI schema to expose everything easily... But in fact after testing there are just 2 or 3 basic endpoints that have been created despite the article/video announcement.

**Comment by KminekMatej** [comment]
@sneko I'm no longer pushing this PR to update Prestashop API. Latest response I got from core team members was that they would need to write autotests for API before this gets merged to ensure stability after upgrade. But no-one wrote them I guess. I gave up already

**Pull request closed without merging**
